code/espurna/sensors/EZOPHSensor.h

```diff
@@ -0,0 +1,210 @@
+
+            if (_serial) delete _serial;
+
+            _serial = new SoftwareSerial(_pin_rx, _pin_tx);
+            _serial->enableIntTx(false);
+            _serial->begin(9600);
+
+            _ready = true;
+            _dirty = false;
+        }
+
+        // Descriptive name of the sensor
+        String description() {
+            char buffer[28];
+            snprintf(buffer, sizeof(buffer), "EZOPH @ SwSerial(%u,%u)", _pin_rx, _pin_tx);
+            return String(buffer);
+        }
+
+        // Descriptive name of the slot # index
+        String slot(unsigned char index) {
+            return description();
+        };
+
+        // Address of the sensor (it could be the GPIO or I2C address)
+        String address(unsigned char index) {
+            char buffer[6];
+            snprintf(buffer, sizeof(buffer), "%u:%u", _pin_rx, _pin_tx);
+            return String(buffer);
+        }
+
+        // Type for slot # index
+        unsigned char type(unsigned char index) {
+            if (index == 0) return MAGNITUDE_PH;
+            return MAGNITUDE_NONE;
+        }
+
+        void tick() {
+            _setup();
+            _read();
+        }
+
+        // Current value for slot # index
+        double value(unsigned char index) {
+            if (index == 0) return _ph;
+            return 0;
+        }
+
+
+
+    protected:
+
+        // ---------------------------------------------------------------------
+        // Protected
+        // ---------------------------------------------------------------------
+
+        void _setup() {
+          if (_sync_responded == true) {
+            return;
+          }
+
+          _error = SENSOR_ERROR_WARM_UP;
+
+          String sync_serial = "";
+          sync_serial.reserve(30);
+
+          if (_sync_requested == false) {
+              _serial->write(67); // C
+              _serial->write(44); // ,
+              _serial->write(63); // ?
+              _serial->write(13); // \r
+
+              _sync_requested = true;
+          }
+
+          while ((_serial->available() > 0)) {
+              char sync_char = (char)_serial->read();
+              sync_serial += sync_char;
+
+              if (sync_char == '\r') {
+                break;
+              }
+          }
+
+          if (sync_serial.startsWith("?C,")) {
+              _sync_interval = sync_serial.substring(sync_serial.indexOf(",") + 1).toInt() * 1000;
+
+              if (_sync_interval == 0) {
+                  _error = SENSOR_ERROR_OTHER;
+                  return;
+              }
+          }
+
+          if (sync_serial.startsWith("*OK")) {
+              _sync_responded = true;
+          }
+
+          if (_sync_responded == false) {
+            return;
+          }
+
+          _error = SENSOR_ERROR_OK;
+        }
+
+        void _read() {
+            if (_error != SENSOR_ERROR_OK) {
+              return;
+            }
+
+            if (millis() - ts <= _sync_interval) {
```

Comment:
This is harmless right now, but isn't it comparing `millis() <= _sync_interval` before first read? _setup should set it initially.

Reply:
It shouldn't because I have `if (_error != SENSOR_ERROR_OK) {` just before that.